hashtree_abi.nim

```diff
@@ -0,0 +1,29 @@
+# hashtree nim bindings
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed and distributed under either of
+#   * MIT license (license terms in the root directory or at https://opensource.org/licenses/MIT).
+#   * Apache v2 license (license terms in the root directory or at https://www.apache.org/licenses/LICENSE-2.0).
+# at your option. This file may not be copied, modified, or distributed except according to those terms.
+
+import std/[os, strutils]
+
+const srcDir = currentSourcePath.parentDir.replace('\\', '/') & "/src/"
+
+{.compile: srcDir & "hashtree.c".}
+
+when defined(arm64):
+  {.compile: srcDir & "sha256_armv8_neon_x1.S".}
+  {.compile: srcDir & "sha256_armv8_neon_x4.S".}
+
+elif defined(amd64):
+  {.compile: srcDir & "sha256_avx_x1.S".}
+  {.compile: srcDir & "sha256_avx_x4.S".}
+  {.compile: srcDir & "sha256_avx_x8.S".}
+  {.compile: srcDir & "sha256_avx_x16.S".}
+  {.compile: srcDir & "sha256_shani.S".}
+  {.compile: srcDir & "sha256_sse_x1.S".}
```

Comment:
Fails for me on macOS Intel.

```sh
% sysctl hw.model
hw.model: MacBookPro16,1
```

```sh
% sw_vers
ProductName:		macOS
ProductVersion:		14.3.1
BuildVersion:		23D60
```

```zsh
% nim c -r tests/hashtree.nim
Hint: used config file '/usr/local/Cellar/nim/2.0.2/nim/config/nim.cfg' [Conf]
Hint: used config file '/usr/local/Cellar/nim/2.0.2/nim/config/config.nims' [Conf]
...................................................................................................................
CC: sha256_avx_x1
CC: sha256_avx_x4
CC: sha256_avx_x8
CC: sha256_avx_x16
CC: sha256_shani
CC: sha256_sse_x1
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:27:17: error: unexpected token in '.section' directive
.section .rodata
                ^
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:112:1: error: unknown directive
.type hashtree_sha256_shani_x2,%function
^
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:130:2: error: invalid operand for instruction
 movdqa SHUF_MASK, [rip + .LPSHUFFLE_BYTE_FLIP_MASK]
 ^
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:131:2: error: invalid operand for instruction
 lea SHA256CONSTANTS,[rip + .LK256]
 ^
/Users/etan/Documents/Repos/hashtree/src/sha256_shani.S:132:2: error: invalid operand for instruction
 lea SHA256PADDING,[rip + .LPADDING]
 ^
... (many more lines)
```

Reply:
this is expected, the library does not yet support mac